subprojects/model-core/src/main/java/org/gradle/internal/instantiation/generator/AbstractClassGenerator.java

```diff
@@ -1147,42 +1145,25 @@ boolean hasNameProperty() {
         }
     }
 
-    private static class PropertyTypePropertyHandler extends ClassGenerationHandler {
-        private final List<PropertyMetadata> propertyTyped = new ArrayList<>();
+    private static class LazyGroovySupportTypePropertyHandler extends ClassGenerationHandler {
+        private final List<PropertyMetadata> lazyGroovySupportTyped = new ArrayList<>();
 
         @Override
         void visitProperty(PropertyMetadata property) {
-            if (property.isReadable() && isPropertyType(property.getType())) {
-                propertyTyped.add(property);
+            if (property.isReadable() && isPropertyType(property.getType())
+                || property.isReadOnly() && isConfigurableFileCollectionType(property.getType())) {
+                lazyGroovySupportTyped.add(property);
             }
         }
 
         @Override
         void applyTo(ClassGenerationVisitor visitor) {
-            for (PropertyMetadata property : propertyTyped) {
+            for (PropertyMetadata property : lazyGroovySupportTyped) {
                 visitor.addPropertySetterOverloads(property, property.mainGetter);
```

Comment:
🤔 Should this be called now `addLazyGroovySupportSetterOverloads()` or something that is not specific about "property"?